core/src/lib/channel.h

```diff
@@ -0,0 +1,349 @@
+
+   This program is Free Software; you can redistribute it and/or
+   modify it under the terms of version three of the GNU Affero General Public
+   License as published by the Free Software Foundation and included
+   in the file LICENSE.
+
+   This program is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301, USA.
+*/
+
+#ifndef BAREOS_LIB_CHANNEL_H_
+#define BAREOS_LIB_CHANNEL_H_
+
+#include <condition_variable>
+#include <optional>
+#include <vector>
+#include <utility>
+#include <variant>
+
+#include "lib/thread_util.h"
+
+namespace channel {
+// a simple single consumer/ single producer queue
+// Its composed of three parts: the input, the output and the
+// actual queue data.
+// Instead of directly interacting with the queue itself you instead
+// interact with either the input or the output.
+// This ensures that there is only one producer (who writes to the input)
+// and one consumer (who reads from the output).
+
+struct failed_to_acquire_lock {};
+struct channel_closed {};
+
+template <typename T> class queue {
+  struct internal {
+    std::vector<T> data;
+    bool in_dead;
+    bool out_dead;
+  };
+
+  synchronized<internal> shared{};
+  std::condition_variable in_update{};
+  std::condition_variable out_update{};
+  std::size_t max_size;
+
+ public:
+  queue(std::size_t max_size) : max_size(max_size) {}
+  queue(const queue&) = delete;
+  queue& operator=(const queue&) = delete;
+  queue(queue&&) = delete;
+  queue& operator=(queue&&) = delete;
+
+  using locked_type = decltype(shared.lock());
+
+  class handle {
+    locked_type locked;
+    std::condition_variable* update;
+
+   public:
+    handle(locked_type locked, std::condition_variable* update)
+        : locked{std::move(locked)}, update(update)
+    {
+    }
+    handle(const handle&) = delete;
+    handle& operator=(const handle&) = delete;
+    handle(handle&&) = delete;
+    handle& operator=(handle&&) = delete;
+
+    std::vector<T>& data() { return locked->data; }
+
+    ~handle() { update->notify_one(); }
+  };
+
+  /* the *_lock functions return std::nullopt only if the channel is closed
+   * otherwise they wait until they get the lock
+   * The try_*_lock functions instead return a tristate indicating whether
+   * they succeeded, failed to acquire the lock or if the channel was closed
+   * from the other side. */
+
+
+  std::optional<handle> read_lock()
+  {
+    auto locked = shared.lock();
+    if (locked->out_dead) {
+      // note(ssura): This happening is programmer error.
+      // Maybe we should assert this instead ?
```

Comment:
A lot of the time the code involving channel necessitate the close method.  For example sometimes I have to write something like this:
```C++
void fun()
{
  auto [in, out] = channel::CreateBufferedChannel(50);
  std::thread consumer{consume, std::move(out)};
  for (;;) {
    std::optional val = produce();
    if (val) {
      in.emplace(std::move(val).value());
    } else {
      goto bail_out;
    }
  }
bail_out:
  in.close();
  consumer.join();
}
```
While this is technically doable without close, it can quickly lead to hard to read code.